src/dune_rules/context.ml

```diff
@@ -355,67 +366,43 @@ type instance =
 let create ~(kind : Kind.t) ~path ~env ~env_nodes ~name ~merlin ~targets
     ~host_context ~host_toolchain ~profile ~fdo_target_exe
     ~dynamically_linked_foreign_archives ~instrument_with =
-  let prog_not_found_in_path prog =
-    Utils.program_not_found prog ~context:name ~loc:None
-  in
   let which = Program.which ~path in
-  let which_exn x =
-    which x >>| function
-    | None -> prog_not_found_in_path x
-    | Some x -> x
-  in
-  let findlib_config_path =
-    Memo.lazy_ ~cutoff:Path.External.equal (fun () ->
-        let* fn = which_exn "ocamlfind" in
-        (* When OCAMLFIND_CONF is set, "ocamlfind printconf" does print the
-           contents of the variable, but "ocamlfind printconf conf" still prints
-           the configuration file set at the configuration time of ocamlfind,
-           sigh... *)
-        (match Env.get env "OCAMLFIND_CONF" with
-        | Some s -> Memo.return s
-        | None ->
-          Memo.of_reproducible_fiber
-            (Process.run_capture_line ~display:Quiet ~env Strict fn
-               [ "printconf"; "conf" ]))
-        >>| Path.External.of_filename_relative_to_initial_cwd)
+  let env_ocamlpath = Findlib.Config.ocamlpath env in
+  let ocamlpath =
+    let initial_ocamlpath = Findlib.Config.ocamlpath Env.initial in
+    match (env_ocamlpath, initial_ocamlpath) with
```

Comment:
Treating the empty and unset OCAMLPATH as unset for this match looks suspicious to me tbh.

Could you add some comments regarding the selection of these rules (and perhaps preserve some of the old ones). This code is confusing enough to deserve some explnations.

Reply:
This is a bit cargo-culted from #1707 though I did check the logic. I'll revert it back to option types which shouldn't change the previous code, and will make it clearer.

Reply:
I made this a lot clearer in [302b45c](https://github.com/ocaml/dune/pull/7428/commits/302b45c68596c7d83bca09df4db27b5ca8fff8c7). Added comments explaining the different branches too.